presto-main/src/main/java/com/facebook/presto/ttl/fetcherManager/ConfidenceBasedTTLFetcherManager.java

```diff
@@ -0,0 +1,191 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static com.facebook.airlift.concurrent.Threads.threadsNamed;
+import static com.facebook.presto.util.PropertiesUtil.loadProperties;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Sets.difference;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class ConfidenceBasedTTLFetcherManager
+        implements TTLFetcherManager
+{
+    private static final Logger log = Logger.get(ConfidenceBasedTTLFetcherManager.class);
+    private static final File TTL_FETCHER_CONFIG = new File("etc/ttl-fetcher.properties");
+    private static final String TTL_FETCHER_PROPERTY_NAME = "ttl-fetcher.factory";
+    private final AtomicReference<TTLFetcher> ttlFetcher = new AtomicReference<>();
+    private final InternalNodeManager nodeManager;
+    private final ConcurrentHashMap<InternalNode, NodeTTL> nodeTTLMap = new ConcurrentHashMap<>();
+    private final boolean isWorkScheduledOnCoordinator;
+    private final Map<String, TTLFetcherFactory> ttlFetcherFactories = new ConcurrentHashMap<>();
+    private final TTLFetcherManagerConfig ttlFetcherManagerConfig;
+    private final Consumer<AllNodes> nodeChangeListener = this::refreshTTLInfo;
+    private ScheduledExecutorService refreshTTLExecutor;
+
+    @Inject
+    public ConfidenceBasedTTLFetcherManager(InternalNodeManager nodeManager, NodeSchedulerConfig schedulerConfig, TTLFetcherManagerConfig ttlFetcherManagerConfig)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.isWorkScheduledOnCoordinator = requireNonNull(schedulerConfig, "nodeSchedulerConfig is null").isIncludeCoordinator();
+        this.ttlFetcherManagerConfig = requireNonNull(ttlFetcherManagerConfig, "ttlFetcherManagerConfig is null");
+    }
+
+    public void scheduleRefresh()
+    {
+        refreshTTLExecutor = newSingleThreadScheduledExecutor(threadsNamed("refresh-ttl-executor-%s"));
+        refreshTTLExecutor.scheduleWithFixedDelay(() -> {
```

Comment:
I would recommend a few things:

1) Scheduling with fixed delay here means it will kick off at a specific interval, no matter how long the task runs for.  I believe that the TTL service this calls may block, so instead probably what we should do is schedule the next execution only after the first one succeeds.  An easy way to ensure this behavior is use the `PeriodicTaskExecutor` utility (which also jitters the calls).
2) We should probably add some metrics here to show freshness of the data, since it's used for scheduling decisions.

Reply:
1. `scheduleWithFixedDelay` does account for how long the task runs. From the docs:

> Creates and executes a periodic action that becomes enabled first after the given initial delay, and subsequently with the given delay between the termination of one execution and the commencement of the next.

However, it doesn't support jitter so I'll modify it to use the `PeriodicTaskExecutor` utility.

2. Great point, let me add something for that.

Reply:
Switched to using `PeriodicTaskExecutor` and added metrics to show freshness of data.